From bbb87fa7c0beaf334694625cdebdbe5b22566f4d Mon Sep 17 00:00:00 2001
From: Jani Hautakangas <jani.hautakangas@lge.com>
Date: Fri, 7 Nov 2014 17:59:03 -0800
Subject: [PATCH] QFontEngine glyph cache count

:Release Notes:
QFontEngine glyph cache limit

:Detailed Notes:
Expose environment variable to control the number of glyph caches used
by QFontEngine.
QT_MAX_GLYPH_CACHE_COUNT
 - By default Qt allows 4 glyph caches per font. By exporting this
   variable in WebAppMgr.conf we can control the number of caches.

:Testing Performed:
export QT_MAX_GLYPH_CACHE_COUNT=1
restart WebAppMgr
launch enyo2sampler

Upstream-Status: Inappropriate [specific to tv]
---
 src/gui/text/qfontengine.cpp | 11 ++++++++---
 src/gui/text/qfontengine_p.h |  1 +
 2 files changed, 9 insertions(+), 3 deletions(-)

diff --git a/src/gui/text/qfontengine.cpp b/src/gui/text/qfontengine.cpp
index 790dd0b..92ea8e0 100644
--- a/src/gui/text/qfontengine.cpp
+++ b/src/gui/text/qfontengine.cpp
@@ -243,7 +243,8 @@ QFontEngine::QFontEngine(Type type)
       font_(0), font_destroy_func(0),
       face_(0), face_destroy_func(0),
       m_minLeftBearing(kBearingNotInitialized),
-      m_minRightBearing(kBearingNotInitialized)
+      m_minRightBearing(kBearingNotInitialized),
+      m_maxGlyphCacheCount(4)
 {
     faceData.user_data = this;
     faceData.get_font_table = qt_get_font_table_default;
@@ -260,6 +261,10 @@ QFontEngine::QFontEngine(Type type)
     if (enginesCollector)
         enginesCollector->append(this);
 #endif
+
+    QByteArray env = qgetenv("QT_MAX_GLYPH_CACHE_COUNT");
+    if (!env.isEmpty())
+        m_maxGlyphCacheCount = env.toInt();
 }
 
 QFontEngine::~QFontEngine()
@@ -1062,9 +1067,9 @@ void QFontEngine::setGlyphCache(const void *context, QFontEngineGlyphCache *cach
             return;
     }
 
-    // Limit the glyph caches to 4 per context. This covers all 90 degree rotations,
+    // Limit the glyph caches to m_maxGlyphCacheCount per context. This covers all 90 degree rotations,
     // and limits memory use when there is continuous or random rotation
-    if (caches.size() == 4)
+    if (caches.size() == m_maxGlyphCacheCount)
         caches.removeLast();
 
     GlyphCacheEntry entry;
diff --git a/src/gui/text/qfontengine_p.h b/src/gui/text/qfontengine_p.h
index 893dfb5..96bd601 100644
--- a/src/gui/text/qfontengine_p.h
+++ b/src/gui/text/qfontengine_p.h
@@ -329,6 +329,7 @@ private:
     mutable qreal m_minLeftBearing;
     mutable qreal m_minRightBearing;
 
+    int m_maxGlyphCacheCount;
 };
 
 Q_DECLARE_OPERATORS_FOR_FLAGS(QFontEngine::ShaperFlags)
